pkg/reconciler/tenancy/initialization/apibinder_initializer_controller.go

```diff
@@ -0,0 +1,311 @@
+			return apiExportsInformer.Lister().Get(clusters.ToClusterAwareKey(clusterName, name))
+		},
+
+		commit: committer.NewCommitter[*tenancyv1alpha1.ClusterWorkspace, *tenancyv1alpha1.ClusterWorkspaceSpec, *tenancyv1alpha1.ClusterWorkspaceStatus](kcpClusterClient.TenancyV1alpha1().ClusterWorkspaces()),
+	}
+
+	c.transitiveTypeResolver = admission.NewTransitiveTypeResolver(c.getClusterWorkspaceType)
+
+	logger := logging.WithReconciler(klog.Background(), ControllerName)
+
+	clusterWorkspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj, logger)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj, logger)
+		},
+	})
+
+	apiBindingsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIBinding(obj, logger)
+
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueueAPIBinding(obj, logger)
+		},
+	})
+
+	clusterWorkspaceTypeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspaceType(obj, logger)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueueClusterWorkspaceType(obj, logger)
+		},
+	})
+
+	return c, nil
+}
+
+type clusterWorkspaceResource = committer.Resource[*tenancyv1alpha1.ClusterWorkspaceSpec, *tenancyv1alpha1.ClusterWorkspaceStatus]
+
+// APIBinder is a controller which instantiates APIBindings and waits for them to be fully bound
+// in new ClusterWorkspaces.
+type APIBinder struct {
+	queue workqueue.RateLimitingInterface
+
+	getClusterWorkspace     func(clusterName logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	getClusterWorkspaceType func(clusterName logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspaceType, error)
+	listClusterWorkspaces   func() ([]*tenancyv1alpha1.ClusterWorkspace, error)
+
+	listAPIBindings  func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error)
+	getAPIBinding    func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIBinding, error)
+	createAPIBinding func(ctx context.Context, clusterName logicalcluster.Name, binding *apisv1alpha1.APIBinding) (*apisv1alpha1.APIBinding, error)
+	deleteAPIBinding func(ctx context.Context, clusterName logicalcluster.Name, bindingName string) error
+
+	getAPIExport func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+
+	transitiveTypeResolver transitiveTypeResolver
+
+	// commit creates a patch and submits it, if needed.
+	commit func(ctx context.Context, new, old *clusterWorkspaceResource) error
+}
+
+type transitiveTypeResolver interface {
+	Resolve(t *tenancyv1alpha1.ClusterWorkspaceType) ([]*tenancyv1alpha1.ClusterWorkspaceType, error)
+}
+
+func (b *APIBinder) enqueueClusterWorkspace(obj interface{}, logger logr.Logger) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```

Comment:
Do you want to fix these up to be `kcpcache` + cluster-aware or run the generator on this?

Reply:
Will fix